File: pyserver/schema.py
"""Field schema and display slots.

Participants have no fixed shape. The
columns come from whatever file the organiser uploads, and the board decides
what to paint from display slots that reference those columns by key.

Mirrors ``server/schema.js`` exactly so both runtimes read the same file.
"""

import math
import re
from typing import Any, Dict, List, Optional, Sequence

MAX_FIELDS = 30
KEY_PATTERN = re.compile(r"^[a-z0-9_]{1,40}$")

EMPHASIS = ("primary", "secondary", "meta", "eyebrow")
SLOTS = ("reel", "call", "card", "panel")

MAX_LINES_PER_SLOT = 6
PANEL_ENTRY_LIMITS = (1, 500)

# Columns that usually carry personal contact details. Only used to pre-tick
# "sensitive" on import — the organiser stays in control of the final answer.
SENSITIVE_HINTS = re.compile(
    r"mobile|phone|contact|email|whatsapp|passport|emirates|id_number|address", re.I
)

# Columns that usually identify the physical ticket, tried in order. A
# sensitive column is never chosen automatically — a phone number happens to be
# unique, but it is not the thing being drawn.
IDENTIFIER_HINTS = (
    re.compile(r"ticket"),
    re.compile(r"badge|entry|registration|booking|coupon|invoice|voucher"),
    re.compile(r"^id$|_id$|^no$|_no$|_number$|_code$|_ref$|serial"),
)


def to_key(label: Any, fallback_index: int = 0) -> str:
    """Turns an arbitrary column heading into a stable, safe object key."""
    slug = re.sub(r"[^a-z0-9]+", "_", str(label or "").strip().lower())
    slug = slug.strip("_")[:40]
    return slug if KEY_PATTERN.match(slug) else f"field_{fallback_index + 1}"


def to_label(key: str) -> str:
    return " ".join(word.capitalize() for word in str(key).split("_") if word)


def normalize_field(raw: Any, index: int) -> Dict[str, Any]:
    source = raw if isinstance(raw, dict) else {"key": raw}
    candidate = source.get("key")
    key = candidate if isinstance(candidate, str) and KEY_PATTERN.match(candidate) else to_key(
        candidate or source.get("label"), index
    )

    label = source.get("label")
    sensitive = source.get("sensitive")

    return {
        "key": key,
        "label": label.strip()[:60] if isinstance(label, str) and label.strip() else to_label(key),
        "sensitive": sensitive is True or (sensitive is None and bool(SENSITIVE_HINTS.search(key))),
        "includeInExport": source.get("includeInExport") is not False,
    }


def normalize_fields(raw: Any) -> List[Dict[str, Any]]:
    items = raw if isinstance(raw, list) else []
    seen = set()
    fields: List[Dict[str, Any]] = []

    for index, entry in enumerate(items):
        field = normalize_field(entry, index)
        if field["key"] in seen or len(fields) >= MAX_FIELDS:
            continue
        seen.add(field["key"])
        fields.append(field)

    return fields


def pick_identifier(fields: Sequence[Dict[str, Any]], preferred: Optional[str]) -> str:
    keys = [field["key"] for field in fields]
    if preferred and preferred in keys:
        return preferred

    safe_keys = [field["key"] for field in fields if not field["sensitive"]]
    for pattern in IDENTIFIER_HINTS:
        match = next((key for key in safe_keys if pattern.search(key)), None)
        if match:
            return match

    if safe_keys:
        return safe_keys[0]
    return keys[0] if keys else "ticket"


def schema_from_columns(columns: Sequence[Any], previous: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Builds a schema from the column headings of an uploaded file."""
    previous = previous or {}
    previous_by_key = {field["key"]: field for field in (previous.get("fields") or [])}

    drafts = []
    for index, column in enumerate(columns):
        key = to_key(column, index)
        prior = previous_by_key.get(key, {})
        # Keep the organiser's own label and sensitivity choices across
        # re-imports; otherwise use the heading exactly as the file wrote it.
        draft = dict(prior)
        draft["key"] = key
        draft["label"] = prior.get("label") or (str(column).strip() or to_label(key))
        drafts.append(draft)

    fields = normalize_fields(drafts)
    return {"identifier": pick_identifier(fields, previous.get("identifier")), "fields": fields}


def _normalize_line(raw: Any, valid_keys: Sequence[str], index: int) -> Optional[Dict[str, Any]]:
    source = raw if isinstance(raw, dict) else {"field": raw}
    if source.get("field") not in valid_keys:
        return None

    emphasis = source.get("emphasis")
    return {
        "field": source["field"],
        "emphasis": emphasis if emphasis in EMPHASIS else ("primary" if index == 0 else "meta"),
        "showLabel": source.get("showLabel") is True,
    }


def _normalize_lines(raw: Any, valid_keys: Sequence[str]) -> List[Dict[str, Any]]:
    items = raw if isinstance(raw, list) else []
    seen = set()
    lines: List[Dict[str, Any]] = []

    for index, entry in enumerate(items):
        line = _normalize_line(entry, valid_keys, index)
        if not line or line["field"] in seen or len(lines) >= MAX_LINES_PER_SLOT:
            continue
        seen.add(line["field"])
        lines.append(line)

    return lines


def default_display(schema: Dict[str, Any]) -> Dict[str, Any]:
    """Sensible slots for a freshly imported schema, used when none are configured."""
    identifier = schema["identifier"]
    fields = schema["fields"]
    others = [field for field in fields if field["key"] != identifier]
    visible = [field for field in others if not field["sensitive"]]
    headline = visible[0] if visible else None
    rest = visible[1:]

    def line(key: Optional[str], emphasis: str) -> List[Dict[str, Any]]:
        return [{"field": key, "emphasis": emphasis, "showLabel": False}] if key else []

    return {
        "reel": {"lines": line(identifier, "primary")},
        "call": {"lines": line(identifier, "primary")},
        "card": {
            "lines": line(headline["key"] if headline else None, "primary")
            + line(identifier, "secondary")
            + [{"field": f["key"], "emphasis": "meta", "showLabel": False} for f in rest[:2]]
        },
        "panel": {
            "lines": line(identifier, "eyebrow")
            + line(headline["key"] if headline else None, "primary")
            + [{"field": f["key"], "emphasis": "meta", "showLabel": False} for f in rest[:1]],
            "maxEntries": 50,
        },
    }


def _clamp_entries(value: Any, fallback: int) -> int:
    if isinstance(value, (list, dict)) or value is None:
        return fallback
    try:
        number = float(value)
    except (TypeError, ValueError):
        return fallback
    if number != number or number in (float("inf"), float("-inf")):
        return fallback
    low, high = PANEL_ENTRY_LIMITS
    # Rounded the way JavaScript rounds, so a half lands on the same integer.
    return int(min(high, max(low, math.floor(number + 0.5))))


def normalize_display(raw: Any, schema: Dict[str, Any]) -> Dict[str, Any]:
    source = raw if isinstance(raw, dict) else {}
    valid_keys = [field["key"] for field in schema["fields"]]
    defaults = default_display(schema)

    display: Dict[str, Any] = {}
    for slot in SLOTS:
        configured = _normalize_lines((source.get(slot) or {}).get("lines"), valid_keys)
        # An empty slot would leave a blank screen on stage, so fall back.
        display[slot] = {"lines": configured or defaults[slot]["lines"]}

    display["panel"]["maxEntries"] = _clamp_entries(
        (source.get("panel") or {}).get("maxEntries"), defaults["panel"]["maxEntries"]
    )
    return display


def public_field_keys(display: Dict[str, Any]) -> List[str]:
    """Every field key the public board is allowed to receive."""
    keys: List[str] = []
    for slot in SLOTS:
        for line in display[slot]["lines"]:
            if line["field"] not in keys:
                keys.append(line["field"])
    return keys


def project_record(record: Dict[str, Any], keys: Sequence[str]) -> Dict[str, Any]:
    return {key: record[key] for key in keys if key in record}

File: server/schema.js
'use strict';

/**
 * Field schema and display slots.
 *
 * Participants have no fixed shape. The
 * columns come from whatever file the organiser uploads, and the board decides
 * what to paint from display slots that reference those columns by key.
 */

const MAX_FIELDS = 30;
const KEY_PATTERN = /^[a-z0-9_]{1,40}$/;

const EMPHASIS = Object.freeze(['primary', 'secondary', 'meta', 'eyebrow']);
const SLOTS = Object.freeze(['reel', 'call', 'card', 'panel']);

const MAX_LINES_PER_SLOT = 6;
const PANEL_ENTRY_LIMITS = Object.freeze([1, 500]);

/** Turns an arbitrary column heading into a stable, safe object key. */
function toKey(label, fallbackIndex = 0) {
  const slug = String(label || '')
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')
    .slice(0, 40);

  return KEY_PATTERN.test(slug) ? slug : `field_${fallbackIndex + 1}`;
}

function toLabel(key) {
  return String(key)
    .split('_')
    .filter(Boolean)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}

// Columns that usually carry personal contact details. Only used to pre-tick
// "sensitive" on import — the organiser stays in control of the final answer.
const SENSITIVE_HINTS = /mobile|phone|contact|email|whatsapp|passport|emirates|id_number|address/i;

// Columns that usually identify the physical ticket, tried in order. A
// sensitive column is never chosen automatically — a phone number happens to
// be unique, but it is not the thing being drawn.
const IDENTIFIER_HINTS = [
  /ticket/,
  /badge|entry|registration|booking|coupon|invoice|voucher/,
  /^id$|_id$|^no$|_no$|_number$|_code$|_ref$|serial/,
];

function normalizeField(input, index) {
  const source = input && typeof input === 'object' ? input : { key: input };
  const key = KEY_PATTERN.test(source.key) ? source.key : toKey(source.key || source.label, index);

  return {
    key,
    label: typeof source.label === 'string' && source.label.trim() ? source.label.trim().slice(0, 60) : toLabel(key),
    sensitive: source.sensitive === true || (source.sensitive === undefined && SENSITIVE_HINTS.test(key)),
    includeInExport: source.includeInExport !== false,
  };
}

function normalizeFields(input) {
  const list = Array.isArray(input) ? input : [];
  const seen = new Set();
  const fields = [];

  list.forEach((entry, index) => {
    const field = normalizeField(entry, index);
    if (seen.has(field.key) || fields.length >= MAX_FIELDS) return;
    seen.add(field.key);
    fields.push(field);
  });

  return fields;
}

function pickIdentifier(fields, preferred) {
  const keys = fields.map((field) => field.key);
  if (preferred && keys.includes(preferred)) return preferred;

  const safeKeys = fields.filter((field) => !field.sensitive).map((field) => field.key);
  for (const pattern of IDENTIFIER_HINTS) {
    const match = safeKeys.find((key) => pattern.test(key));
    if (match) return match;
  }

  return safeKeys[0] || keys[0] || 'ticket';
}

/** Builds a schema from the column headings of an uploaded file. */
function schemaFromColumns(columns, previous = {}) {
  const previousByKey = new Map((previous.fields || []).map((field) => [field.key, field]));

  const fields = normalizeFields(
    columns.map((column, index) => {
      const key = toKey(column, index);
      const previousField = previousByKey.get(key);
      // Keep the organiser's own label and sensitivity choices across
      // re-imports; otherwise use the heading exactly as the file wrote it.
      return {
        ...(previousField || {}),
        key,
        label: (previousField && previousField.label) || String(column).trim() || toLabel(key),
      };
    })
  );

  return { identifier: pickIdentifier(fields, previous.identifier), fields };
}

function normalizeLine(input, validKeys, index) {
  const source = input && typeof input === 'object' ? input : { field: input };
  if (!validKeys.includes(source.field)) return null;

  return {
    field: source.field,
    emphasis: EMPHASIS.includes(source.emphasis) ? source.emphasis : index === 0 ? 'primary' : 'meta',
    showLabel: source.showLabel === true,
  };
}

function normalizeLines(input, validKeys) {
  const list = Array.isArray(input) ? input : [];
  const seen = new Set();
  const lines = [];

  list.forEach((entry, index) => {
    const line = normalizeLine(entry, validKeys, index);
    if (!line || seen.has(line.field) || lines.length >= MAX_LINES_PER_SLOT) return;
    seen.add(line.field);
    lines.push(line);
  });

  return lines;
}

/** Sensible slots for a freshly imported schema, used when none are configured. */
function defaultDisplay(schema) {
  const { identifier, fields } = schema;
  const others = fields.filter((field) => field.key !== identifier);
  const visible = others.filter((field) => !field.sensitive);
  const [headline, ...rest] = visible;

  const line = (field, emphasis) => (field ? [{ field: field.key, emphasis, showLabel: false }] : []);

  return {
    reel: { lines: line({ key: identifier }, 'primary') },
    call: { lines: line({ key: identifier }, 'primary') },
    card: {
      lines: [
        ...line(headline, 'primary'),
        ...line({ key: identifier }, 'secondary'),
        ...rest.slice(0, 2).map((field) => ({ field: field.key, emphasis: 'meta', showLabel: false })),
      ],
    },
    panel: {
      lines: [
        ...line({ key: identifier }, 'eyebrow'),
        ...line(headline, 'primary'),
        ...rest.slice(0, 1).map((field) => ({ field: field.key, emphasis: 'meta', showLabel: false })),
      ],
      maxEntries: 50,
    },
  };
}

function clampEntries(value, fallback) {
  // FIX: as in settings.js — null, '' and [] are not numbers anybody gave us,
  // and Number() turning them into 0 disagreed with the Python side.
  const numeric =
    typeof value === 'number' ||
    typeof value === 'boolean' ||
    (typeof value === 'string' && value.trim() !== '');
  if (!numeric) return fallback;

  const number = Number(value);
  if (!Number.isFinite(number)) return fallback;
  const [min, max] = PANEL_ENTRY_LIMITS;
  return Math.min(max, Math.max(min, Math.round(number)));
}

function normalizeDisplay(input, schema) {
  const source = input && typeof input === 'object' ? input : {};
  const validKeys = schema.fields.map((field) => field.key);
  const defaults = defaultDisplay(schema);

  const display = SLOTS.reduce((accumulator, slot) => {
    const configured = normalizeLines((source[slot] || {}).lines, validKeys);
    // An empty slot would leave a blank screen on stage, so fall back.
    return { ...accumulator, [slot]: { lines: configured.length > 0 ? configured : defaults[slot].lines } };
  }, {});

  display.panel.maxEntries = clampEntries((source.panel || {}).maxEntries, defaults.panel.maxEntries);
  return display;
}

/** Every field key the public board is allowed to receive. */
function publicFieldKeys(display) {
  return [...new Set(SLOTS.flatMap((slot) => display[slot].lines.map((line) => line.field)))];
}

function projectRecord(record, keys) {
  return keys.reduce((accumulator, key) => {
    if (record[key] === undefined) return accumulator;
    return { ...accumulator, [key]: record[key] };
  }, {});
}

module.exports = {
  EMPHASIS,
  SLOTS,
  MAX_FIELDS,
  MAX_LINES_PER_SLOT,
  toKey,
  toLabel,
  normalizeFields,
  pickIdentifier,
  schemaFromColumns,
  normalizeDisplay,
  defaultDisplay,
  publicFieldKeys,
  projectRecord,
};
